### ISSUE TITLE
ELF file conflict 'color' resolution for tri-lib systems

### ISSUE THREAD

**Issue by mhatle** [rpm-software-management/rpm#193]
I am attempting to use rpm with a tri-lib system, mips64.  (ELF 32, ELF64 and MIPS64 n32)

I have a small patch that adds MIPS64 n32 ABI support to RPM:  http://git.openembedded.org/openembedded-core/tree/meta/recipes-devtools/rpm/files/0001-Add-a-color-setting-for-mips64_n32-binaries.patch

This introduces color value of '4', that indicates MIPS64 n32.

This works if the system is installing MIPS64 n32 vs MIPS64 (colors 4 vs 2) and I have the preferred color set to one or the other.  However, if I add the third library of '1', or MIPS(32) then it can't resolve a three way conflict properly.

I've identified that in the lib/transaction.c file, handleColorConflict function.  The issue is that the system needs a final 'else' clause to deal with the situation where neither is the preferred type.

I've been working on a patch to resolve this, but I can't figure out how to make it work properly:

                rConflicts = 0;
+           } else {
+               /*
+                * If neither is already skipped, we skip the old one, and
+                * install the new one (last in wins).
+                */
+               if (ofs && !XFA_SKIPPING(rpmfsGetAction(ofs, ofx)) &&
+                    fs && !XFA_SKIPPING(rpmfsGetAction(fs, fx))) {
+                   rpmfsSetAction(ofs, ofx, FA_SKIPCOLOR);
+                   rpmfsSetAction(fs, fx, FA_CREATE);
+               }
+               rConflicts = 0;
            }

When I have do an install with ELF32, ELF64 and MIPS64 n32 (specified in that order as the rpm transaction), setting the preferred color to '2' (ELF64), the system will still install the n32 version.  (If I change the order so MIPS64 n32 is not the last in the transaction order then it works fine.)

What I appear to be getting during processing is:

conflict -- MIPS64 n32 vs ELF32 -- resolved to ELF32 (via the else I added)
conflict -- ELF32 vs ELF64 -- resolved to ELF64 (via existing mechanism)
conflict -- ELF64 vs ELF32 -- resolved to ELF64 (via existing mechanism)
conflict -- ELF32 vs MIPS64 n32 -- resolves to MIPS64 n32 (via the else I added)

So the conflict appears resolved, and both ELF64 and MIPS64n32 are installed, with the n32 version being 'last', so that is the version on the disk.

Any help would be appreciated on this three way resolution.

**Comment by n3npq** [comment]
Resolving file "colors"  affects more than MIPS64 n32: the x32 format also uses (1 <<2) color bit and (potentially) needs RPM to choose between executables that are compiled differently.

AFAIK, PLD has succeeded in adding support for x32 to RPM5, if you need patches that have been tested in the "real world".

**Comment by mhatle** [comment]
Yes x32 has exactly the same issue.  The RPM5 code for the resolver is very different then this code set.  So unfortunately it can't be directly applied.

I simply don't understand the conflict resolution loop in this code base enough to understand why I'm not able to do the three-way resolution.
